### ISSUE TITLE
B::Deparse (use Module Version)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6514]
Migrated from [rt.perl.org#22216](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22216) (status was 'resolved')

Searchable as RT22216$

**Comment by p5pRT** [comment]
### From at@altlinux.ru

#### Created by at@altlinux.ru



Hi\,

I'm trying to use B&#8203;::Deparse for perl\.req script that automatically discovers
dependencies in perl files \(RPM\)\.  Though generally it works well\, there are
but a few defects\.  The following example shows its inability to handle module
version in use statements properly\.

$ perl \-MO=Deparse \-e 'use Socket 1\.3;'
use Socket 1;
;
\-e syntax OK
$ perl \-MO=Deparse \-e 'use Socket 1\.7;'
use Socket 1;
;
\-e syntax OK
$

Thanks\,
Alexey Tourbin
ALT Linux Team

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.0 - Fri Dec 27 20:51:43 MSK 2002
It is being executed now by  Perl v5.8.0 - Mon Dec 30 15:08:38 MSK 2002.

Site configuration information for perl v5.8.0:

Configured by inger at Mon Dec 30 15:08:38 MSK 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.18-alt7-smp, archname=i386-linux-thread-multi
    uname='linux mash.office.altlinux.ru 2.4.18-alt7-smp #1 smp thu jun 6 11:43:56 msd 2002 i686 unknown unknown gnulinux '
    config_args='-des -Darchname=i386-linux -Dd_dosuid -Ud_csh -Duseshrplib -Dlibperl=libperl.so.5.8 -Doptimize=-pipe -Wall -O2 -fexpensive-optimizations -march=i586 -mcpu=i686 -Dinstallprefix=/usr -Dprefix=/usr -Dprivlib=/usr/lib/perl5 -Darchlib=/usr/lib/perl5/i386-linux -Dvendorprefix=/usr -Dvendorlib=/usr/lib/perl5/vendor_perl -Dvendorarch=/usr/lib/perl5/vendor_perl/i386-linux -Dsiteprefix=/usr/local -Dsitelib=/usr/local/lib/perl5/site_perl/5.8.0 -Dsitearch=/usr/local/lib/perl5/site_perl/5.8.0/i386-linux -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dcf_by=ALT Linux Team -Dmyhostname=localhost -Dperladmin=root@localhost -Dinc_version_list=5.6.1 5.6.0  -Dpager=/usr/bin/less -isR -Di_shadow -Di_syslog -Dusethreads -Duseithreads -Duselargefiles -Ui_db -Di_gdbm -Di_ndbm -Di_sdbm -Ui_odbm'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-pipe -Wall -O2 -fexpensive-optimizations -march=i586 -mcpu=i686',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -I/usr/include/gdbm'
    ccversion='', gccversion='3.2.1 (ALT Linux, build 3.2.1-alt2)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.6.so, so=so, useshrplib=true, libperl=libperl.so.5.8
    gnulibc_version='2.2.6'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/lib/perl5/i386-linux
    /usr/lib/perl5
    /usr/local/lib/perl5/site_perl/5.8.0/i386-linux
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/i386-linux
    /usr/lib/perl5/vendor_perl
    /usr/lib/perl5/vendor_perl
    .


Environment for perl v5.8.0:
    HOME=/home/at
    LANG=C
    LANGUAGE=C
    LC_COLLATE=ru_RU.CP1251
    LC_CTYPE=ru_RU.CP1251
    LC_MESSAGES=C
    LC_MONETARY=ru_RU.CP1251
    LC_NUMERIC=ru_RU.CP1251
    LC_TIME=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/at/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/local/bin:/usr/games:/usr/lib/rpm:/sbin:/usr/sbin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Alexey Tourbin \(via RT\) \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> $ perl \-MO=Deparse \-e 'use Socket 1\.3;'
> use Socket 1;

Something like this should handle this case\, and the v\-string
case as well :

Index&#8203;: Deparse\.pm
===================================================================
\-\-\- Deparse\.pm  \(revision 1156\)
\+\+\+ Deparse\.pm  \(working copy\)
@&#8203;@&#8203; \-326\,7 \+326\,10 @&#8203;@&#8203;
 
&nbsp;        return unless $self\->const\_sv\($constop\)\->PV eq $module;
&nbsp;        $constop = $constop\->sibling;
\-       $version = $self\->const\_sv\($constop\)\->int\_value;
\+       $version = $self\->const\_sv\($constop\)\->PV;
\+       if \($version =~ /\[^0\-9\.\]/\) \{
\+           $version = 'v'\.join '\.'\, map ord\, split //\, $version;
\+       \}
&nbsp;        $constop = $constop\->sibling;
&nbsp;        return if $constop\->name ne "method\_named";
&nbsp;        return if $self\->const\_sv\($constop\)\->PV ne "VERSION";

**Comment by p5pRT** [comment]
### From @rgs

Alexey Tourbin \(via RT\) wrote&#8203;:
> $ perl \-MO=Deparse \-e 'use Socket 1\.3;'
> use Socket 1;

I applied this patch : \(thanks to Spider Boardman for suggesting
to look at the type of the const\_sv\)

Change 19539 by rgs@&#8203;rgs\-home on 2003/05/17 11&#8203;:36&#8203;:08

&nbsp;        Fix bug \#22216 : B&#8203;::Deparse can't handle "use Module Version"
&nbsp;        with fractional version numbers or v\-strings\.

Affected files \.\.\.

\.\.\. //depot/perl/ext/B/B/Deparse\.pm\#129 edit

Differences \.\.\.

==== //depot/perl/ext/B/B/Deparse\.pm\#129 \(text\) ====

@&#8203;@&#8203; \-326\,7 \+326\,14 @&#8203;@&#8203;
 
&nbsp;        return unless $self\->const\_sv\($constop\)\->PV eq $module;
&nbsp;        $constop = $constop\->sibling;
\-       $version = $self\->const\_sv\($constop\)\->int\_value;
\+       $version = $self\->const\_sv\($constop\);
\+       if \(class\($version\) ne "PVMG"\) \{
\+           \# version is either an integer or a double
\+           $version = $version\->PV;
\+       \} else \{
\+           \# version specified as a v\-string
\+           $version = 'v'\.join '\.'\, map ord\, split //\, $version\->PV;
\+       \}
&nbsp;        $constop = $constop\->sibling;
&nbsp;        return if $constop\->name ne "method\_named";
&nbsp;        return if $self\->const\_sv\($constop\)\->PV ne "VERSION";
End of Patch\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
